lib/rubocop/cop/style/pattern_argument.rb

```diff
@@ -0,0 +1,122 @@
+      #
+      # @example
+      #   # bad
+      #   %w[foo bar foobar].any? { |e| /oo/.match?(e) }
+      #
+      #   [1, 2, 3].any? { |e| (1..10).include?(e) }
+      #
+      #   [1, 2, 3].any? { |e| Set[1, 2, 3] === e }
+      #
+      #   [1, 2, 3].none? do |e|
+      #     Integer === e
+      #   end
+      #
+      #   # good
+      #   %w[foo bar foobar].any?(/oo/)
+      #
+      #   [1, 2, 3].any?(1..10)
+      #
+      #   [1, 2, 3].any?(Set[1, 2, 3])
+      #
+      #   [1, 2, 3].none?(Integer)
+      #
+      class PatternArgument < Cop
+        include RangeHelp
+        extend TargetRubyVersion
+
+        minimum_target_ruby_version 2.5
+
+        MSG = 'Pass `%<pattern>s` as an argument to `%<method>s` instead of a block.'
+
+        def_node_matcher :pattern_as_parameter_candidate?, <<~PATTERN
+          (block $({send csend} _ {:all? :any? :none? :one?}) $_ $_)
+        PATTERN
+
+        def on_block(node)
+          pattern_as_parameter_candidate?(node) do |method, args, body|
+            receiver = body.receiver
+            return unless offense?(args, body, receiver)
+
+            add_offense(node, location: method.loc.selector,
+                              message: format(MSG, method: method.method_name,
+                                                   pattern: pattern(args, body, receiver).source))
+          end
+        end
+
+        def autocorrect(node)
+          lambda do |corrector|
+            body = node.body
+            return unless regexp?(body, body.receiver) || body.method?(:is_a?) ||
+                          case_equality_alias?(body.method_name)
+            return unless
+              (replacement = PatternArguments::PatternExtractor.new(node).pattern.source)
+
+            replacement = "(#{replacement})" unless replacement.match?(/^\(.+\)$/)
+            corrector.replace(block_range_with_space(node), replacement)
+          end
+        end
+
+        private
+
+        def offense?(args, body, receiver)
+          return unless args.size == 1
+          return unless body.respond_to?(:method?)
+
+          case_equality?(args, body) || body.method?(:is_a?) || supported_obj?(body, receiver)
+        end
+
+        def case_equality?(args, body)
+          body.method?(:===) && args.child_nodes.first.source != body.receiver.source
+        end
+
+        def supported_obj?(body, receiver)
+          range?(body, receiver) || regexp?(body, receiver) || set?(body, receiver)
+        end
+
+        def range?(body, receiver)
+          obj, = *receiver
+          receiver.range_type? || (receiver.begin_type? && obj.range_type? &&
+            case_equality_alias?(body.method_name))
+        end
+
+        def regexp?(body, receiver)
+          obj, = *receiver
+          _, obj_class = *obj
+          (receiver.regexp_type? || obj_class == :Regexp) &&
+            %i[=== =~ match match?].include?(body.method_name)
+        end
+
+        def set?(body, receiver)
```

Comment:
Have you considered using NodePattern for these kind of checks? It would probably be more readable.

Also it looks like `set?` is missing some logic, as `:Set` does not appear in the source code...

Reply:
Thanks @marcandre, the truth is I just followed some other cops while working on this one. I updated it to use the NodePattern and I got rid of all those checks, amazing.

I'm not sure if there's a better way to get [this](https://github.com/rubocop-hq/rubocop/pull/8220/files#diff-c0aa68a59ced803500d8a79422f5ee7fR48) done.